src/plugins/mer/meremulatormodeoptionswidget.cpp

```diff
@@ -0,0 +1,233 @@
+    QTC_ASSERT(!m_deviceModels.value(name).isSdkProvided(), return false);
+
+    m_deviceModels.remove(name);
+
+    m_ui->deviceModelComboBox->setDeviceModels(m_deviceModels.values());
+    m_ui->deviceModelComboBox->setCurrentIndex(0);
+    m_ui->optionsWidget->setDeviceModelsNames(m_deviceModels.uniqueKeys());
+
+    return true;
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelNameChanged(const QString &newName)
+{
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    renameDeviceModel(name, newName);
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelDconfChanged(const QString &value)
+{
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.dconf() != value) {
+        updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                 model.displayResolution(),
+                                                 model.displaySize(),
+                                                 value,
+                                                 model.isSdkProvided()));
+    }
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelScreenResolutionChanged(const QSize &resolution)
+{
+    const auto name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.displayResolution() != resolution) {
+        updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                 resolution,
+                                                 model.displaySize(),
+                                                 model.dconf(),
+                                                 model.isSdkProvided()));
+    }
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelScreenSizeChanged(const QSize &size)
+{
+    const auto name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.displaySize() != size) {
+        updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                 model.displayResolution(),
+                                                 size,
+                                                 model.dconf(),
+                                                 model.isSdkProvided()));
+    }
+}
+
+void MerEmulatorModeOptionsWidget::updateGui()
+{
+    if (m_ui->deviceModelComboBox->count() == 0)
+        return;
+
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    if (m_deviceModels.keys().contains(name)) {
+        const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+        m_ui->removeProfileButton->setDisabled(model.isSdkProvided());
+        m_ui->optionsWidget->setCurrentDeviceModel(model);
+    }
```

Comment:
If you ask me: I use asserts for some uncommon, unexpected, but possible situation. This situation is common - when new emulator device added the emulator mode is empty.

Reply:
Sounds like a bug - an emulator should always have its mode defined. It was never discovered because currently you cannot remove/add the emulator back inside Options. So I would add a commit after "Move reading SDK provided device models to MerSettings", to fix it. The commit would initialize MerEmulatorDevice::deviceModel with the first model in the list (we currently do not have a notion of "default" device model.). Doing that inside MerDeviceFactory::create() seems to be the right way.